### ISSUE TITLE
iCloud encryption

### ISSUE THREAD

**Issue by BenjaminMichaelis** [twofas/2fas-ios#43]
I would love it if the backups to iCloud for iOS could be client side encrypted as I believe by default it is not e2ee encrypted by advanced data protection even, and also support (https://2fas.com/support/security-privacy/is-2fas-backup-safe/) says there is an ability to set a password and I am not seeing that option on iOS for iCloud backup (only exporting a file I see)

**Comment by adocyn** [comment]
@BenjaminMichaelis Hi, yes, currently on iOS only the `export to file` backup is encrypted by custom password. The iCloud backup is encrypted, but using a constant key. We have a plan to add this functionality in future but unfortunately can't get exact timeline for that. If you enable Advanced Data Protection it will additionally encrypt the CloudKit database used by 2FAS https://support.apple.com/en-gb/guide/security/sec973254c5f/web

**Comment by BenjaminMichaelis** [comment]
> @BenjaminMichaelis Hi, yes, currently on iOS only the `export to file` backup is encrypted by custom password. The iCloud backup is encrypted, but using a constant key. We have a plan to add this functionality in future but unfortunately can't get exact timeline for that. If you enable Advanced Data Protection it will additionally encrypt the CloudKit database used by 2FAS https://support.apple.com/en-gb/guide/security/sec973254c5f/web

It does use Advanced Data Protection? I don't see the 'encryptedValues' property used anywhere?

**Comment by adocyn** [comment]
@BenjaminMichaelis You're right, general docs states that you have to "only enable this feature" but fields you've mentioned are in docs and probably used for that exact purpose (the docs are terrible). We'll take a look into that.

**Comment by valynor** [comment]
Could we get an update/comment on this from the devs, please?

**Comment by BenjaminMichaelis** [comment]
Any update on this @adocyn ?

**Comment by adocyn** [comment]
Queued for 5.4. Currently 5.3 is in the works.

**Comment by DannieBGoode** [comment]
+1 for this functionality

**Comment by elliotwutingfeng** [comment]
Is there a technical restriction on the iOS platform that prevents the use of encryption libraries on the client-side to encrypt backups before uploading to iCloud?

**Comment by DannieBGoode** [comment]
> Is there a technical restriction on the iOS platform that prevents the use of encryption libraries on the client-side to encrypt backups before uploading to iCloud?

There isn't, this is how Raivo does it and they don't seem to have any problem with it.

**Comment by Penguinjumper** [comment]
> Queued for 5.4. Currently 5.3 is in the works.

very excited for this. Might even make the recommendation on Privacyguides possible. (https://discuss.privacyguides.net/t/add-2fas-authenticator-app/12958/56)

**Comment by bhall7** [comment]
@DannieBGoode, yes! The ability to add a user password like Raivo would make me feel a little better about iCloud syncing. 🤞

**Comment by huyz** [comment]
When is 5.4 coming out?  Just wanted to note that this issue is a blocker for my even trying 2FAS out.
At the very least, Advanced Data Protection should work.

**Comment by DjKilla2** [comment]
It's been a while since there's been an update on this issue. Is this still being planned for the 5.4 update? I too have migrated recently from Authy to Ente Auth which has E2EE encryption and is recommended by both Privacy Guides and Techlore. My understanding is that once this issue is fixed, 2FAS will join the elite ranks of both Privacy Guides and Techlore which many apps have tried to be a part of but few have made it. I'm still hoping this fix is being worked on as this is what is keeping me from using 2FAS as well. Anyway, hope to hear an update soon!

**Comment by bvwpo** [comment]
@huyz 
> When is 5.4 coming out? Just wanted to note that this issue is a blocker for my even trying 2FAS out. At the very least, Advanced Data Protection should work.

FYI, iCloud sync option is off by default. So you can try 2FAS now without worrying about this ADP issue, since all your data will be stored locally on your phone. Only if you turn on backup/sync it will use iCloud, and that's when ADP is needed.

**Comment by huyz** [comment]
> So you can try 2FAS now without worrying about this ADP issue, since all your data will be stored locally on your phone.

Thanks for that, but I still wouldn't try a solution without encrypted backups because I have no idea how long I would have to wait to have a feasible solution.

**Comment by bvwpo** [comment]
> Thanks for that, but I still wouldn't try a solution without encrypted backups because I have no idea how long I would have to wait to have a feasible solution.

I’m on the same boat; only testing it with a few fake TOTP’s now. I’m waiting for this feature to be implemented before moving my real accounts over.

**Comment by githubusername600** [comment]
Any update on this? This is keeping me from using 2fas as well.

**Comment by adocyn** [comment]
Hi, no updates yet. We're experimenting with different solutions because such change is a major redesign of the iCloud storage. So it should be done once and correctly 🙂

**Comment by Goldmaster** [comment]
> @BenjaminMichaelis Hi, yes, currently on iOS only the `export to file` backup is encrypted by custom password. The iCloud backup is encrypted, but using a constant key. We have a plan to add this functionality in future but unfortunately can't get exact timeline for that. If you enable Advanced Data Protection it will additionally encrypt the CloudKit database used by 2FAS https://support.apple.com/en-gb/guide/security/sec973254c5f/web

This does need to be fixed as 2fas was rejected from privacyguides due to lack of icloud encryption

https://discuss.privacyguides.net/t/add-2fas-authenticator-app/12958

Once implemented, may be best contacting them directly to get 2fas added.

**Comment by tavern2782** [comment]
A year without a fix. So much for a privacy focused app eh?

**Comment by SethuSenthil** [comment]
I think this is an essential feature. If there is anything I could do to contribute I would love to.

**Comment by tavern2782** [comment]
I've asked for updates on Reddit, here, and Discord. The 2FAS team refuses to provide an ETA as to when 5.4 is releasing.

**Comment by RadCod3** [comment]
Been following this issue for a while and I switched to Ente auth due to this. 2fas still looks and _feels_ a lot better than ente auth tho so I hope this gets implemented at some point in the future. Till then I'll use something that has e2e encryption.

**Comment by chrisbchrisb** [comment]
It's over a year now. I am also considering moving to Ente. 
Are there any updates in sight about this issue?

**Comment by huyz** [comment]
> When is 5.4 coming out? Just wanted to note that this issue is a blocker for my even trying 2FAS out. At the very least, Advanced Data Protection should work.

Yeah it's been a while with no communication. Giving up and unsubscribing…

**Comment by timkgh** [comment]
> I've asked for updates on Reddit, here, and Discord. The 2FAS team refuses to provide an ETA as to when 5.4 is releasing.

Clear sign to look elsewhere. This is not how security apps should be managed.

**Comment by valynor** [comment]
> It's over a year now. I am also considering moving to Ente. Are there any updates in sight about this issue?

I have to agree with this. I know it's open source and free and everything but this is also a security relevant app so I think we deserve some update after almost half a year. 

@adocyn

**Comment by tavern2782** [comment]
Hey don't worry everyone, some nobody user on Discord is mad that we're rightfully mad. He said we just should use a VPN or trusted network when backing up with iCloud. When I pointed out his poor understanding of how iCloud works, he said just disable the backup until you're on a VPN or trusted network.

Why didn't we think of that? The glorious user Mr. Wiz just saved us all!

**Comment by valynor** [comment]
I waited another 2 weeks, hopeful for a life sign from the 2FAS devs alas ... nothing. 
Even though this is an open source app and I appreciate the work of the devs this is not what I want to see from an important app like a 2nd factor authenticator. 

I'm off to Ente Auth.

**Comment by tavern2782** [comment]
Just want to point out that @adocyn has been making tons of commits to private repos since January but won't comment on 2FAS. Must hate his users.

**Comment by autom4il** [comment]
This is an open source software, for free. The developers surely worked hard to make it possible for the users to use 2FA, they don't make money with it unless from donations. I absolutely agree to the fact that E2EE must be implemented, doesn't mean that for them its easy and it eventually it will be implemented. For anyone, make pull request if you know how to code and what to do and see that probably isn't that easy as you think. There are other alternatives out there for the impatient.

**Comment by tavern2782** [comment]
> This is an open source software, for free. The developers surely worked hard to make it possible for the users to use 2FA, they don't make money with it unless from donations. I absolutely agree to the fact that E2EE must be implemented, doesn't mean that for them its easy and it eventually will be implemented. For anyone, make pull request if you know how to code and what to do and see that probably isn't that easy as you think. There are other alternatives out there for the ones that wants E2EE as soon as possible.

No one cares what you think.

**Comment by Flupsy** [comment]
> This is an open source software, for free. The developers surely worked hard to make it possible for the users to use 2FA, they don't make money with it unless from donations. I absolutely agree to the fact that E2EE must be implemented, doesn't mean that for them its easy and it eventually will be implemented. For anyone, make pull request if you know how to code and what to do and see that probably isn't that easy as you think. There are other alternatives out there for the ones that wants E2EE as soon as possible.

Well said. If this was my project and I got this many toxic comments, I’d say ‘why am I doing this to myself?’ and just walk away. I’m sure the dev(s) here are better people than I am, and will stick with it, but they are volunteers. 

If it’s important to you as a user to have someone to shout at* when you’re unhappy, go buy a commercial product with a support contract. 

* of course you should never shout at support staff either!

**Comment by adocyn** [comment]
Hi guys. Doing my best. A lot is happening right now, but version 5.4 with iCloud encryption is in the works. Now. Like I'm working on it this minute. Yes, I know, it's Sunday 😄 I can only tell you that it's not easy. And yes, I'm working on other things too hence the private repos and, well, commercial work. You know how it is. If I want to use my computer, I need to have my electric bill payed on time 🙂 About @tavern2782 - sorry if you're disappointed, but personal comments like that don't help. What helps is donations and maybe pull requests. That's all. Anyway, back to code 😄

**Comment by Goldmaster** [comment]
Hey dude,

Sorry to see you get a bad comment like that.

As someone who is the ideas person and not a programmer. I always say *never* rush a dev and I would just wait it or donate. Reminds me of a youtube video awesome open source made on another user who commented something in a similar way https://youtu.be/8vxuiOvfRP8

Take it easy @adocyn and don't rush. It's better to get something out working first time than release a stable version that has bugs or such issues.

Look after yourself.

**Comment by tavern2782** [comment]
All we wanted was an update since this has been pending over a year. We got one thanks to me. So you can all be mad about how I approached it but I got results. You’re welcome.

**Comment by adriancchase** [comment]
I've been considering switching over to Ente Auth but I'm hyped to first see if E2EE update will come out soon. 2FAS is definitely cleaner and I prefer having the data stored on iCloud vs Ente's servers.

Just to clear the confusion, does ADP encrypt your secrets, yes or no? This was addressed in this issue but I couldn't understand clearly if ADP helps at all with 2FAS backups?

Furthermore, respect for the dev who does this in their free time. Thanks!

**Comment by jonaharagon** [comment]
ADP does **not** currently protect your data, unless the developer sets their app up to do so. That is the crux of this issue here.

**Comment by dahs94** [comment]
Can anyone help me understand how it works at the moment in more detail? It's not my area, so it would take me ages to figure it out by looking through the repo. 

The developer above says that the data in iCloud is encrypted using a constant key. Am I right in thinking that this means a key controlled & accessible to 2FAS is used to encrypt the data during backup, and then when that data reaches iCloud, it's encrypted at again by Apple's standard data protection? So to get to this data at rest, someone would need the apple key and 2FAS constant key?

Really like the app, just doing some due diligence before it's used in my organisation. 

Thanks in advance.

**Comment by DannieBGoode** [comment]
> Can anyone help me understand how it works at the moment in more detail? It's not my area, so it would take me ages to figure it out by looking through the repo.
> 
> The developer above says that the data in iCloud is encrypted using a constant key. Am I right in thinking that this means a key controlled & accessible to 2FAS is used to encrypt the data stored in iCloud, which is encrypted again using Apple's standard data protection? So to get to this data at rest, someone would need the apple key and 2FAS constant key?
> 
> Really like the app, just doing some due diligence before it's used in my organisation.
> 
> Thanks in advance.

The Backup that is stored in iCloud is not encrypted by 2FAS (or by the user) in any way, so the only encryption it is subjected to is the one that iCloud provides.

This is usually fine, as long as you keep your iCloud backup safe (specially if you turn on ADP), nobody should be able to access your 2FAS Backup.

However most of us are asking for an additional encryption done on the backup before storing the file in iCloud. This is supported by the 2FAS for backups exported as a file, but for some reason not for iCloud backups.

**Comment by BenjaminMichaelis** [comment]
> > Can anyone help me understand how it works at the moment in more detail? It's not my area, so it would take me ages to figure it out by looking through the repo.
> > 
> > The developer above says that the data in iCloud is encrypted using a constant key. Am I right in thinking that this means a key controlled & accessible to 2FAS is used to encrypt the data stored in iCloud, which is encrypted again using Apple's standard data protection? So to get to this data at rest, someone would need the apple key and 2FAS constant key?
> > 
> > Really like the app, just doing some due diligence before it's used in my organisation.
> > 
> > Thanks in advance.
> 
> The Backup that is stored in iCloud is not encrypted by 2FAS (or by the user) in any way, so the only encryption it is subjected to is the one that iCloud provides.
> 
> This is usually fine, as long as you keep your iCloud backup safe (specially if you turn on ADP), nobody should be able to access your 2FAS Backup.
> 
> However most of us are asking for an additional encryption done on the backup before storing the file in iCloud. This is supported by the 2FAS for backups exported as a file, but for some reason not for iCloud backups.


To clarify, one of the issues is that it does not currently use the correct APIs in order to actually utilize ADP from our look, so it is only protected by normal iCloud protection.

**Comment by adocyn** [comment]
Check the branch 5.4.0. We're still testing as this version has to migrate the data and well, we don't want to screw this up 😅

**Comment by chrisbchrisb** [comment]
Looking forward to the 5.4 release. For those who don't want to wait. Proton just released an end2end encrypted stand allone authenticator app. https://proton.me/authenticator

**Comment by adocyn** [comment]
Branch for 5.4.0 created. There were major changes to the Vault structure so we can allow users with older versions keep using the sync while moving users of the new version to much stronger Vault. It wasn't easy, but this should be the final version.

**Comment by Chess07** [comment]
How is it going with this issue? When are you going to release 5.4?

**Comment by hahahahacker2009** [comment]
> How is it going with this issue? When are you going to release 5.4?

2 more years

**Comment by huyz** [comment]
Yes it's over. This ship has sailed. Many repo owners are busy with AI.
And now we have passkeys. 
Move on too

**Comment by adocyn** [comment]
We are in the final stabilization phase for 5.4, and we're genuinely close to the release. This update touches the core key handling (incl. key encoding and migration paths), so we are finishing extended regression, migration and security tests. It is taking this long, as we want to ensure it ships correctly and safely for our users.

**Comment by adocyn** [comment]
5.4.0 has been published. Sorry for the long wait. Busy time.

**Comment by rubnogueira** [comment]
@adocyn my iCloud storage is full currently, and I just opened the app, and it's stuck in the "Migration Cloud backup" screen for at least 5 minutes.

I wonder if that edge case was handled, because I feel that is taking too long.

Edit: I tried closing and opening the app, and I get stuck in that screen. I can't use any of the codes (only in Airplane mode).

**Comment by ac9492** [comment]
Mine doesn't seem to sync correctly across my two devices. I updated in one and it syncs, but the second one just gets this error. When I tap "Ok" it simply disables sync and stays that way. I had ADP enabled before I updated, don't know if that has anything to do. I also had iCloud Keychain disabled, but I enabled it on both devices but still same error.



![image](https://github.com/user-attachments/assets/577a26dc-a4ba-4fc9-8807-2dfd047cab85)

EDIT: I see the ability to set a custom password on my first device, but I haven't enabled it. No luck trying to resolve this.

EDIT #2: I seem to have solved this. I had to delete the cloud backup and then try syncing again while both devices had keychain on. It seems to have happened because I had keychain off initially and the backup won't sync with the new encryption even if you enable sync after the fact.

**Comment by dpantel** [comment]
My app does not have the option to update yet. Are you guys on the newest version?

**Comment by adocyn** [comment]
> I can't use any of the codes (only in Airplane mode).
@rubnogueira

Hi, it could be full but it doesn't report that to the app. We've tested such scenarios. iOS may try to remove unnecessary content before reporting that it can't save and that the iCloud account is full. If it starts normally in Airplane Mode, then for now, please disable the sync. It won't migrate or send anything anyway until there's enough space in the cloud.

Also, how many services you have in the app?

**Comment by adocyn** [comment]
@dpantel 

> My app does not have the option to update yet. Are you guys on the newest version?

It's a stager rollout but the update should be available for manual updating in the App Store.

**Comment by dpantel** [comment]
@adocyn Thank you for this upgrade. I have been waiting for it to transition to 2FAS. 

Can you please elaborate how this update solves the cloud backup issue?

How are things stored? How they’re encrypted? What the difference is between using a password for the backup or not. 

Thank you.

**Comment by rubnogueira** [comment]
> > I can't use any of the codes (only in Airplane mode).
> > [@rubnogueira](https://github.com/rubnogueira)
> 
> Hi, it could be full but it doesn't report that to the app. We've tested such scenarios. iOS may try to remove unnecessary content before reporting that it can't save and that the iCloud account is full. If it starts normally in Airplane Mode, then for now, please disable the sync. It won't migrate or send anything anyway until there's enough space in the cloud.
> 
> Also, how many services you have in the app?

@adocyn Around 30. I removed some things from iCloud, and as soon as it didn't mark 100% usage, I was able to sync and set the encryption password. So the issue is definitely related to that. I recommend having some "skip" if that happens or a proper validation.

On my second device, I could access my tokens with v5.4.0, but I didn't get that prompt to migrate the backup. It was syncing, but I didn't have the option to set my password, so certainly I was not using the new DB. 

I uninstalled the app, removed it from my "iCloud" backup, and reinstalled it again. After that, I received a prompt to put in my password and everything was sorted.

Regarding the watchOS app, it is completely broken. Crashes when I try to open, same as #204.

**Comment by adocyn** [comment]
@rubnogueira it's possible that if the iCloud wasn't fully synced to the second device. Reinstallation did prompt refetching the data.

**Comment by adocyn** [comment]
@rubnogueira aaaaaand I have the crash. Investigating 🙂

**Comment by adocyn** [comment]
@ac9492 Yes, we had to store the "system key" somewhere and we didn't want to be a static one embedded inside the app as it wouldn't made any sense.

@dpantel We wanted to secure user data while also making the app accessible to non-technical users. To achieve this, we developed two types of encryption, both based on AES-GCM.
The first type uses a key generated by the app. It is unique to every app instance and is securely stored in Apple Keychain. This is the default option and all existing users will be migrated to it automatically, with no change to their current experience. The second type derives the key from a user-provided password. The key is also cached in Keychain, so if the user reinstalls the app, they will need to provide the password or wipe the iCloud Backup contents. Both mechanisms require iCloud Keychain Sync and support seamless syncing across devices.
Encrypted iCloud Sync is also available on watchOS. However, if a custom password is used, the watch must be paired with the iOS device by scanning a QR code. Once paired, the watch is authorized to access the iCloud Backup contents.
We also support Apple's Advanced Data Protection for iCloud (ADP) in this version

**Comment by Staubgeborener** [comment]
> [@adocyn](https://github.com/adocyn) my iCloud storage is full currently, and I just opened the app, and it's stuck in the "Migration Cloud backup" screen for at least 5 minutes.
> 
> I wonder if that edge case was handled, because I feel that is taking too long.
> 
> Edit: I tried closing and opening the app, and I get stuck in that screen. I can't use any of the codes (only in Airplane mode).

I can confirm this. My iCloud storage is full and i get the migration message on my iPhone. If i close the app and open it up again, i still get the message. Which means i can't access my 2FA codes anymore at this step, only if i completely disable mobile data on my iPhone app settings. Even re-installing the app won't help.

**Comment by adocyn** [comment]
I've added improved detection of the 'iCloud storage full' error in version 5.4.1, which is currently pending Apple review. 🤞

**Comment by adocyn** [comment]
It's live. If anybody still has a full iCloud drive, it would be great if you could run the migration and check if it detects it correctly.

**Comment by Staubgeborener** [comment]
The message changed from "Migration Cloud backup" to "Improvements are in progress. Please wait." (to be more clear, i am using the german version of 2FAS so its "Verbesserung im Gange Bitte warten..."). I waited 10 minutes, the screen doesn't changed.

From the end user's perspective, therefore, only the message has changed.

**Comment by adocyn** [comment]
OK, thank you. The detection still doesn't handle the "iCloud is full" error correctly or iCloud isn't returning that error at all. We tested that scenario and it did work, but it seems that if you have backups, caches, etc., iCloud tries to remove unnecessary data first, and that takes time. It would be better if it failed immediately. @Staubgeborener, are you by any chance on our Discord? I could send you a QR code that would upload logs which probably contain the error iCloud is returning.

**Comment by Staubgeborener** [comment]
@adocyn i sent you a private message on discord

**Comment by adocyn** [comment]
Better handling of full iCloud account is in v5.4.2.

**Comment by Staubgeborener** [comment]
> Better handling of full iCloud account is in v5.4.2.

It's working fine

**Comment by adocyn** [comment]
@Staubgeborener Thanks for all the help!

**Comment by toyo2333** [comment]
Is the "user-provided password" in "Backup Management" the same as the password in the "Export" feature?
